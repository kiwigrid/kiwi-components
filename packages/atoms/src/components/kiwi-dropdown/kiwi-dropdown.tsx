import {
  Component,
  ComponentInterface,
  Event,
  EventEmitter,
  h,
  JSX,
  Listen,
  Prop,
  State,
} from '@stencil/core';

@Component({
  tag: 'kiwi-dropdown',
  shadow: false,
})
export class KiwiDropdown implements ComponentInterface {
  private dropdownMenu: HTMLDivElement | undefined;

  /**
   * Css class to be applied to container.
   */
  @Prop()
  containerClass!: string;

  /** Type of the toggle button. */
  @Prop()
  toggleButtonType: 'default' | 'primary' | 'danger' | 'warning' | 'info' =
    'default';

  @State()
  private _open = false;

  private hover = true;

  /**
   * Event signaling this dropdown is being closed.
   */
  @Event()
  closeDropdown!: EventEmitter<void>;

  @Listen('click', { target: 'window' })
  handleOutsideClick(): void {
    if (!this.hover) {
      this.open = false;
    }
  }

  render(): JSX.Element {
    return (
      <div
        class={`btn-group dropdown ${this.open ? 'open' : ''} ${
          this.containerClass
        }`}
        onMouseEnter={this.handleMenuEnter}
        onMouseLeave={this.handleMenuLeave}
      >
        <button
          class={`btn btn-${this.toggleButtonType} dropdown-toggle`}
          onClick={this.handleDropdownToggle}
        >
          <slot name="dropdown-toggle" />
        </button>
        <div
          class="dropdown-menu"
          ref={(el) => (this.dropdownMenu = el as HTMLDivElement)}
        >
          <slot name="dropdown-content" />
        </div>
      </div>
    );
  }

  private handleDropdownToggle: () => void = () => {
    this.open = !this.open;
  };
  private handleMenuEnter: () => void = () => (this.hover = true);
  private handleMenuLeave: () => void = () => (this.hover = false);

  componentDidRender(): void {
    this.adjustPositionAtScreenEdge();
  }

  /**
   * Make sure that the dropdown does not stick over the right edge of the body
   */
  private adjustPositionAtScreenEdge(): void {
    if (this.dropdownMenu) {
      if (this.open) {
        const rect = this.dropdownMenu.getBoundingClientRect();

        if (rect.right > window.visualViewport.width) {
          const offset =
            rect.width > window.visualViewport.width
              ? Math.floor(rect.left)
              : Math.ceil(rect.right - window.visualViewport.width);

          this.dropdownMenu.style.translate = `-${offset}px`;
        }
      } else {
        this.dropdownMenu.style.translate = '';
      }
    }
  }

  private set open(isOpen: boolean) {
    this._open = isOpen;

    if (!isOpen) {
      this.closeDropdown.emit();
    }
  }

  private get open(): boolean {
    return this._open;
  }
}
