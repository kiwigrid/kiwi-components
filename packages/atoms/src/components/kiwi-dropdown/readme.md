# kiwi-dropdown



<!-- Auto Generated Below -->


## Properties

| Property              | Attribute                | Description                                                                                             | Type                                                        | Default     |
| --------------------- | ------------------------ | ------------------------------------------------------------------------------------------------------- | ----------------------------------------------------------- | ----------- |
| `closeOnContentClick` | `close-on-content-click` | Setting this to true will cause the dropdown to close if a click is registered inside the dropdown-menu | `boolean`                                                   | `false`     |
| `containerClass`      | `container-class`        | Css class to be applied to container.                                                                   | `string`                                                    | `''`        |
| `toggleButtonType`    | `toggle-button-type`     | Type of the toggle button.                                                                              | `"danger" \| "default" \| "info" \| "primary" \| "warning"` | `'default'` |


## Events

| Event           | Description                                    | Type                |
| --------------- | ---------------------------------------------- | ------------------- |
| `closeDropdown` | Event signaling this dropdown is being closed. | `CustomEvent<void>` |


## Dependencies

### Used by

 - [kiwi-checkbox-dropdown-menu](../kiwi-checkbox-dropdown-menu)

### Graph
```mermaid
graph TD;
  kiwi-checkbox-dropdown-menu --> kiwi-dropdown
  style kiwi-dropdown fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
