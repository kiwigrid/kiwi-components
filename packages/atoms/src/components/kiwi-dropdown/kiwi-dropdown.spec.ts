import { newSpecPage, SpecPage } from '@stencil/core/testing';
import { KiwiDropdown } from './kiwi-dropdown';
import { expectDefined } from '../util/testing';

const newKiwiDropdown = (): Promise<SpecPage> =>
  newSpecPage({
    components: [KiwiDropdown],
    html: `
        <kiwi-dropdown container-class="fanzy">
          <span slot="dropdown-toggle">Toggle Me</span>
          <span slot="dropdown-content">Boring content</span>
        </kiwi-dropdown>
      `,
  });

describe('kiwi-dropdown', () => {
  beforeEach(() => {
    Object.defineProperty(window, 'visualViewport', {
      value: { width: 1920 },
      writable: true,
    });
  });

  it('renders', async () => {
    const dropdown = await newKiwiDropdown();
    expect(dropdown.root).toMatchSnapshot('initially closed');

    dropdown.root
      ?.querySelector<HTMLButtonElement>('.dropdown-toggle')
      ?.click();
    await dropdown.waitForChanges();

    expect(dropdown.root).toMatchSnapshot('opened on click');
  });

  it.each`
    left    | right   | expected
    ${50}   | ${350}  | ${''}
    ${1700} | ${2000} | ${'-80px'}
    ${100}  | ${3000} | ${'-100px'}
  `(
    'translates the dropdown { left: $left, right: $right } horizontally by $expected for FullHD screen',
    async ({ left, right, expected }) => {
      const dropdown = await newKiwiDropdown();

      const dropdownMenu = dropdown.root?.querySelector(
        '.dropdown-menu',
      ) as HTMLDivElement;
      expectDefined(dropdownMenu);

      dropdownMenu.getBoundingClientRect = jest.fn(() => {
        return {
          left: left,
          right: right,
          width: right - left,
        } as DOMRect;
      });

      dropdown.root
        ?.querySelector<HTMLButtonElement>('.dropdown-toggle')
        ?.click();
      await dropdown.waitForChanges();

      expect(dropdownMenu.style.translate).toEqual(expected);
    },
  );

  it('closes on outside click', async () => {
    const dropdown = await newKiwiDropdown();

    dropdown.root
      ?.querySelector<HTMLButtonElement>('.dropdown-toggle')
      ?.click();
    await dropdown.waitForChanges();

    expect(
      dropdown.root?.querySelector('.dropdown')?.classList.contains('open'),
    ).toBe(true);

    const dropdownDiv = dropdown.root?.querySelector('.dropdown');
    expect(dropdownDiv).toBeTruthy();

    if (dropdownDiv) {
      dropdownDiv.dispatchEvent(new MouseEvent('mouseenter'));
      dropdown.win.dispatchEvent(new MouseEvent('click'));
      await dropdown.waitForChanges();

      expect(dropdownDiv.classList.contains('open')).toBe(true);

      dropdownDiv.dispatchEvent(new MouseEvent('mouseleave'));
      dropdown.win.dispatchEvent(new MouseEvent('click'));
      await dropdown.waitForChanges();

      expect(dropdownDiv.classList.contains('open')).toBe(false);
    }
  });
});
