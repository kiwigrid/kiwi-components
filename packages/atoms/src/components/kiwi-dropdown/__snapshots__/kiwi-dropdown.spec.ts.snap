// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kiwi-dropdown renders: initially closed 1`] = `
<kiwi-dropdown container-class="fanzy">
  <!---->
  <div class="btn-group dropdown fanzy">
    <button class="btn btn-default dropdown-toggle">
      <span slot="dropdown-toggle">
        Toggle Me
      </span>
    </button>
    <div class="dropdown-menu">
      <span slot="dropdown-content">
        Boring content
      </span>
    </div>
  </div>
</kiwi-dropdown>
`;

exports[`kiwi-dropdown renders: opened on click 1`] = `
<kiwi-dropdown container-class="fanzy">
  <!---->
  <div class="btn-group dropdown fanzy open">
    <button class="btn btn-default dropdown-toggle">
      <span slot="dropdown-toggle">
        Toggle Me
      </span>
    </button>
    <div class="dropdown-menu">
      <span slot="dropdown-content">
        Boring content
      </span>
    </div>
  </div>
</kiwi-dropdown>
`;
